File: packages/python/src/sah_sdk/zyins/isa_options.py
"""Typed options-bag constructor sugar for :class:`Isa`.

Mirrors ``packages/ts/src/zyins/isaOptions.ts``. The historic factories
(:meth:`Isa.with_bearer`, :meth:`Isa.with_keycode`, :meth:`Isa.for_form`,
:meth:`Isa.authenticate`) remain the canonical primitives; this surface
is the recommended path going forward and matches the cross-language
SDK shape (TS ``Isa.create({auth, engine, ...})``, C# ``new Isa(new
IsaOptions { Auth = ... })``)::

    isa = Isa.create(
        auth=BearerAuth.from_token("isa_live_..."),
        engine=RemoteEngine.default(),
        timeout=30.0,
        api_version="v2",
    )

``api_version`` is immutable per-instance options metadata for the
native v2 follow-up. Current Python operations still use the
date-pinned platform ``Version`` header and existing endpoint paths.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from ..core.transport import Transport

# ---------------------------------------------------------------------------
# Pinned API major version
# ---------------------------------------------------------------------------

#: Pinned API major version. Default ``"v2"``; immutable per-instance.
IsaApiVersion = Literal["v1", "v2"]


# ---------------------------------------------------------------------------
# Auth suppliers
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class BearerAuthSupplier:
    """Bearer-token auth supplier — accepted by :class:`IsaCreateOptions`.

    ``token`` may be ``None`` to indicate "resolve from ``ISA_TOKEN``"
    (matches the TS shape: ``BearerAuth.fromEnv()``).
    """

    kind: Literal["bearer"] = "bearer"
    token: str | None = None


@dataclass(frozen=True, slots=True)
class LicenseAuthSupplier:
    """License-credential auth supplier."""

    kind: Literal["license"] = "license"
    keycode: str | None = None
    email: str | None = None


@dataclass(frozen=True, slots=True)
class FormAuthSupplier:
    """Form-token auth supplier (embedded eApp)."""

    kind: Literal["form"] = "form"
    form_token: str = ""


@dataclass(frozen=True, slots=True)
class SessionAuthSupplier:
    """Session-credential auth supplier."""

    kind: Literal["session"] = "session"
    session_id: str | None = None
    session_secret: str | None = None


IsaAuthSupplier = (
    BearerAuthSupplier | LicenseAuthSupplier | FormAuthSupplier | SessionAuthSupplier
)


class BearerAuth:
    """Bearer-auth supplier factory.

    Mirrors the TS ``BearerAuth.fromToken(...)`` / ``fromEnv()`` shape so
    consumers reading both SDKs see the same surface.
    """

    @staticmethod
    def from_token(token: str) -> IsaAuthSupplier:
        """Construct from an explicit token. Validates non-emptiness."""
        if not token:
            raise ValueError("BearerAuth.from_token: token must be a non-empty string")
        return BearerAuthSupplier(token=token)

    @staticmethod
    def from_env() -> IsaAuthSupplier:
        """Construct from ``ISA_TOKEN`` at factory time."""
        return BearerAuthSupplier(token=None)


class LicenseAuth:
    """License-auth supplier factory.

    Mirrors the TS ``LicenseAuth.fromKeycode(...)`` / ``fromEnv()`` shape.
    """

    @staticmethod
    def from_keycode(keycode: str, email: str) -> IsaAuthSupplier:
        """Construct from explicit keycode + email."""
        if not keycode:
            raise ValueError("LicenseAuth.from_keycode: keycode must be non-empty")
        if not email:
            raise ValueError("LicenseAuth.from_keycode: email must be non-empty")
        return LicenseAuthSupplier(keycode=keycode, email=email)

    @staticmethod
    def from_env() -> IsaAuthSupplier:
        """Construct from ``ISA_LICENSE_KEYCODE`` + ``ISA_LICENSE_EMAIL``."""
        return LicenseAuthSupplier(keycode=None, email=None)


class FormAuth:
    """Form-token auth supplier factory."""

    @staticmethod
    def from_token(form_token: str) -> IsaAuthSupplier:
        if not form_token:
            raise ValueError("FormAuth.from_token: form_token must be non-empty")
        return FormAuthSupplier(form_token=form_token)


class SessionAuth:
    """Session-auth supplier factory."""

    @staticmethod
    def from_credentials(session_id: str, session_secret: str) -> IsaAuthSupplier:
        """Construct from explicit session credentials."""
        if not session_id:
            raise ValueError(
                "SessionAuth.from_credentials: session_id must be non-empty"
            )
        if not session_secret:
            raise ValueError(
                "SessionAuth.from_credentials: session_secret must be non-empty"
            )
        return SessionAuthSupplier(
            session_id=session_id,
            session_secret=session_secret,
        )

    @staticmethod
    def from_env() -> IsaAuthSupplier:
        """Construct from ``ISA_SESSION_ID`` + ``ISA_SESSION_SECRET``."""
        return SessionAuthSupplier(session_id=None, session_secret=None)


# ---------------------------------------------------------------------------
# Engine selectors
# ---------------------------------------------------------------------------

#: Production ZyINS endpoint origin.
PRODUCTION_REMOTE_ORIGIN = "https://zyins.isaapi.com"

#: Production proxy endpoint origin.
PRODUCTION_PROXY_ORIGIN = "https://proxy.isaapi.com"


@dataclass(frozen=True, slots=True)
class RemoteEngineSelector:
    """Tagged engine selector — production / staging ZyINS endpoint."""

    kind: Literal["remote"] = "remote"
    base_url: str = PRODUCTION_REMOTE_ORIGIN


@dataclass(frozen=True, slots=True)
class LocalEngineSelector:
    """Local engine binary — points at a developer or test endpoint."""

    kind: Literal["local"] = "local"
    base_url: str = "http://localhost:8080"


@dataclass(frozen=True, slots=True)
class ProxyEngineSelector:
    """Routes through the platform proxy (``/v1/call``)."""

    kind: Literal["proxy"] = "proxy"
    proxy_origin: str = PRODUCTION_PROXY_ORIGIN


@dataclass(frozen=True, slots=True)
class InMemoryEngineSelector:
    """In-process mock — bypasses HTTP entirely. Test-only.

    Carries the test transport via the ``transport`` attribute on
    :class:`IsaCreateOptions`; this selector is the marker.
    """

    kind: Literal["in_memory"] = "in_memory"


IsaEngine = (
    RemoteEngineSelector
    | LocalEngineSelector
    | ProxyEngineSelector
    | InMemoryEngineSelector
)


class RemoteEngine:
    """Production ZyINS endpoint factory."""

    @staticmethod
    def default() -> IsaEngine:
        """Default — production endpoint (``https://zyins.isaapi.com``)."""
        return RemoteEngineSelector(base_url=PRODUCTION_REMOTE_ORIGIN)

    @staticmethod
    def at(base_url: str) -> IsaEngine:
        """Construct from an explicit base URL (staging, region-specific)."""
        if not base_url:
            raise ValueError("RemoteEngine.at: base_url must be a non-empty string")
        return RemoteEngineSelector(base_url=base_url)


class LocalEngine:
    """Local engine factory."""

    @staticmethod
    def at(base_url: str) -> IsaEngine:
        if not base_url:
            raise ValueError("LocalEngine.at: base_url must be a non-empty string")
        return LocalEngineSelector(base_url=base_url)


class ProxyEngine:
    """Platform proxy factory."""

    @staticmethod
    def default() -> IsaEngine:
        return ProxyEngineSelector(proxy_origin=PRODUCTION_PROXY_ORIGIN)

    @staticmethod
    def at(proxy_origin: str) -> IsaEngine:
        if not proxy_origin:
            raise ValueError("ProxyEngine.at: proxy_origin must be a non-empty string")
        return ProxyEngineSelector(proxy_origin=proxy_origin)


#: In-process mock engine — pair with ``transport=`` on
#: :class:`IsaCreateOptions` to inject a test transport.
InMemoryEngine: IsaEngine = InMemoryEngineSelector()


# ---------------------------------------------------------------------------
# Options bag
# ---------------------------------------------------------------------------

#: Default per-call timeout in seconds (matches the TS 30_000 ms default).
DEFAULT_TIMEOUT_SECONDS = 30.0


@dataclass(frozen=True, slots=True)
class IsaCreateOptions:
    """Options accepted by :meth:`Isa.create`.

    Every field is optional except ``auth``; defaults match the
    production posture (RemoteEngine.default, 30s timeout, v2).
    """

    auth: IsaAuthSupplier
    engine: IsaEngine | None = None
    timeout: float = DEFAULT_TIMEOUT_SECONDS
    api_version: IsaApiVersion = "v2"
    client_version: str | None = None
    transport: Transport | None = None


@dataclass(frozen=True, slots=True)
class ResolvedIsaOptions:
    """Resolved view of :class:`IsaCreateOptions` with defaults applied.

    Pure value object — safe to pass between the public ``Isa.create``
    facade and the internal ``ZyInsClient`` constructor.
    """

    auth: IsaAuthSupplier
    engine: IsaEngine
    timeout_seconds: float
    api_version: IsaApiVersion
    client_version: str | None
    transport: Transport | None
    base_url: str
    proxy_origin: str | None


def resolve_isa_options(opts: IsaCreateOptions) -> ResolvedIsaOptions:
    """Resolve :class:`IsaCreateOptions` into a fully-defaulted view.

    Pure — no side effects, safe to call from constructors and tests
    alike. Mirrors the TS ``resolveIsaOptions()`` semantics exactly.
    """
    engine = opts.engine if opts.engine is not None else RemoteEngine.default()
    base_url = _engine_base_url(engine)
    proxy_origin = engine.proxy_origin if isinstance(engine, ProxyEngineSelector) else None
    return ResolvedIsaOptions(
        auth=opts.auth,
        engine=engine,
        timeout_seconds=opts.timeout,
        api_version=opts.api_version,
        client_version=opts.client_version,
        transport=opts.transport,
        base_url=base_url,
        proxy_origin=proxy_origin,
    )


def _engine_base_url(engine: IsaEngine) -> str:
    if isinstance(engine, (RemoteEngineSelector, LocalEngineSelector)):
        return engine.base_url
    if isinstance(engine, ProxyEngineSelector):
        # Proxy mode targets the production remote origin for the
        # underlying ZyINS request; the proxy_origin lives on the
        # ResolvedIsaOptions for the proxy namespace to consume.
        return PRODUCTION_REMOTE_ORIGIN
    if isinstance(engine, InMemoryEngineSelector):
        return PRODUCTION_REMOTE_ORIGIN
    # Defensive — exhaustive over the Union; an unknown variant is a
    # programming error rather than a runtime fallback.
    raise TypeError(f"resolve_isa_options: unknown engine selector {engine!r}")

File: packages/ts/src/zyins/isaOptions.ts
/**
 * Typed options-bag constructor sugar for `Isa`.
 *
 * The historic factories (`Isa.withBearer`, `Isa.withKeycode`,
 * `Isa.forForm`, `Isa.authenticate`) remain the canonical primitives; this
 * surface is the recommended path going forward and matches the
 * cross-language SDK shape (C# `new Isa(new IsaOptions { Auth = … })`).
 *
 * ```ts
 * const isa = await Isa.create({
 *   auth: BearerAuth.fromToken('isa_live_…'),
 *   engine: RemoteEngine.default,
 *   timeout: 30_000,
 *   apiVersion: 'v2',
 * });
 * ```
 *
 * `apiVersion` is immutable per-instance and pinned via the `Api-Version`
 * request header. `'v2'` (default) routes `prequalify` to `/v2/prequalify`
 * and returns the typed offer envelope. `'v1'` preserves the legacy
 * contract for callers still wired against the old envelope.
 */

import type { CredentialStore } from '../core';
import type { LicenseRefreshedListener } from './credentialState';
import { type Transport } from './transport';
import {
  type IsaApiVersion,
  type IsaApiVersionOverride,
  type IsaApiSurface,
  BundledApiVersions,
  resolveApiVersions,
} from './bundledApiVersions';

export type { IsaApiVersion, IsaApiVersionOverride, IsaApiSurface };
export { BundledApiVersions, resolveApiVersions };

/**
 * Tagged auth supplier accepted by {@link IsaCreateOptions.auth}. Each
 * variant carries the material the matching factory requires; a single
 * dispatch in `Isa.create` picks the correct legacy factory.
 */
export type IsaAuthSupplier =
  | { readonly kind: 'bearer'; readonly token?: string }
  | {
      readonly kind: 'license';
      readonly keycode?: string;
      readonly email?: string;
      readonly orderId?: string;
      readonly licenseKey?: string;
      readonly credentialStore?: CredentialStore;
      readonly onLicenseRefreshed?: LicenseRefreshedListener;
    }
  | { readonly kind: 'form'; readonly formToken: string }
  | { readonly kind: 'session'; readonly sessionId: string; readonly sessionSecret: string };

/**
 * Bearer-auth supplier factory. Resolves a server-to-server `isa_live_…`
 * token from the explicit argument or, when omitted, from `ISA_TOKEN`.
 */
export const BearerAuth = {
  /** Construct from an explicit token. */
  fromToken(token: string): IsaAuthSupplier {
    return { kind: 'bearer', token };
  },
  /** Construct from `ISA_TOKEN` at factory time. */
  fromEnv(): IsaAuthSupplier {
    return { kind: 'bearer' };
  },
} as const;

/**
 * License-auth supplier factory. Resolves `(keycode, email)` from the
 * explicit args or, when omitted, from `ISA_LICENSE_KEYCODE` +
 * `ISA_LICENSE_EMAIL`.
 */
export const LicenseAuth = {
  /** Construct from explicit keycode + email. */
  fromKeycode(
    keycode: string,
    email: string,
    extras?: Omit<
      Extract<IsaAuthSupplier, { kind: 'license' }>,
      'kind' | 'keycode' | 'email'
    >,
  ): IsaAuthSupplier {
    return { kind: 'license', keycode, email, ...(extras ?? {}) };
  },
  /** Construct from environment variables at factory time. */
  fromEnv(): IsaAuthSupplier {
    return { kind: 'license' };
  },
} as const;

/** Form-token auth supplier factory (embedded eApp). */
export const FormAuth = {
  fromToken(formToken: string): IsaAuthSupplier {
    return { kind: 'form', formToken };
  },
} as const;

/**
 * Tagged engine selector — every variant exposes the same product API
 * surface so consumer code stays identical when targeting different
 * deployments. Values are interpreted by `Isa.create` to set `baseUrl`
 * (and proxyOrigin where applicable).
 */
export type IsaEngine =
  | { readonly kind: 'remote'; readonly baseUrl: string }
  | { readonly kind: 'local'; readonly baseUrl: string }
  | { readonly kind: 'proxy'; readonly proxyOrigin: string }
  | { readonly kind: 'in_memory'; readonly transport: Transport };

const PRODUCTION_REMOTE_ORIGIN = 'https://zyins.isaapi.com';
const PRODUCTION_PROXY_ORIGIN = 'https://proxy.isaapi.com';
const IN_MEMORY_TRANSPORT_ERROR =
  'InMemoryEngine requires inMemoryEngineWith(transport) before product calls';

/** Production ZyINS endpoint (`https://zyins.isaapi.com`). */
export const RemoteEngine = {
  /** Default — production endpoint. */
  default: { kind: 'remote', baseUrl: PRODUCTION_REMOTE_ORIGIN } as IsaEngine,
  /** Construct from an explicit base URL (staging, region-specific). */
  at(baseUrl: string): IsaEngine {
    return { kind: 'remote', baseUrl };
  },
} as const;

/** Local engine binary — points at a developer or test endpoint. */
export const LocalEngine = {
  at(baseUrl: string): IsaEngine {
    return { kind: 'local', baseUrl };
  },
} as const;

/** Routes through the platform proxy (`/v1/call`). */
export const ProxyEngine = {
  default: { kind: 'proxy', proxyOrigin: PRODUCTION_PROXY_ORIGIN } as IsaEngine,
  at(proxyOrigin: string): IsaEngine {
    return { kind: 'proxy', proxyOrigin };
  },
} as const;

/**
 * In-process mock — bypasses HTTP entirely. Test-only; pass an injected
 * transport to satisfy the same surface used by real engines.
 */
export const InMemoryEngine: IsaEngine = {
  kind: 'in_memory',
  transport: async () => {
    throw new Error(IN_MEMORY_TRANSPORT_ERROR);
  },
};

/**
 * Construct an in-memory engine backed by a custom transport. Useful when
 * tests want to assert on outbound request shape without the default
 * transport's network hop.
 */
export function inMemoryEngineWith(transport: Transport): IsaEngine {
  return { kind: 'in_memory', transport };
}

/** Default per-call timeout in milliseconds. */
export const DEFAULT_TIMEOUT_MS = 30_000;

/**
 * Options accepted by `Isa.create()`. Every field is optional except
 * `auth`; defaults match the production posture.
 */
export interface IsaCreateOptions {
  /** Auth supplier. See {@link BearerAuth}, {@link LicenseAuth}, {@link FormAuth}. */
  auth: IsaAuthSupplier;
  /** Engine selector. Defaults to {@link RemoteEngine.default}. */
  engine?: IsaEngine;
  /** Per-call timeout in ms. Defaults to {@link DEFAULT_TIMEOUT_MS}. */
  timeout?: number;
  /**
   * Per-surface API-version override. Surfaces absent from the override
   * fall back to {@link BundledApiVersions}. There is no shorthand string
   * form and no `default` key.
   */
  apiVersion?: IsaApiVersionOverride;
  /** Viewer origin used to assemble case share links. */
  caseViewerBaseUrl?: string;
  /** Optional consumer build identifier for client-version negotiation. */
  clientVersion?: string;
  /** Transport override applied AFTER engine selection (test only). */
  transport?: Transport;
  /**
   * Optional case-storage adapter. Defaults to
   * {@link import('./cases/ZeroKnowledgeCaseStorage').ZeroKnowledgeCaseStorage}
   * — client-side AES-256-GCM, opaque ciphertext on the wire, fragment
   * key as the recall token. Carrier adapters may substitute a portal-
   * backed store; the SDK only sees the {@link CaseStorage} interface.
   */
  caseStorage?: import('./cases/CaseStorage').CaseStorage;
}

/** Resolved view of `IsaCreateOptions` with defaults applied. */
export interface ResolvedIsaOptions {
  readonly auth: IsaAuthSupplier;
  readonly engine: IsaEngine;
  readonly timeoutMs: number;
  /** Fully-resolved per-surface version map (override overlaid on bundled). */
  readonly apiVersions: Readonly<Record<IsaApiSurface, IsaApiVersion>>;
  readonly caseViewerBaseUrl: string | undefined;
  readonly clientVersion: string | undefined;
  readonly transport: Transport | undefined;
  readonly baseUrl: string;
  readonly proxyOrigin: string | undefined;
  readonly caseStorage: import('./cases/CaseStorage').CaseStorage | undefined;
}

/**
 * Resolve `IsaCreateOptions` into a fully-defaulted view. Pure — no side
 * effects, safe to call from constructors and tests alike.
 */
export function resolveIsaOptions(opts: IsaCreateOptions): ResolvedIsaOptions {
  const engine = opts.engine ?? RemoteEngine.default;
  const baseUrl = engineBaseUrl(engine);
  const proxyOrigin = engine.kind === 'proxy' ? engine.proxyOrigin : undefined;
  const engineTransport = engine.kind === 'in_memory' ? engine.transport : undefined;
  const transport = opts.transport ?? engineTransport;
  return {
    auth: opts.auth,
    engine,
    timeoutMs: opts.timeout ?? DEFAULT_TIMEOUT_MS,
    apiVersions: resolveApiVersions(opts.apiVersion),
    caseViewerBaseUrl: opts.caseViewerBaseUrl,
    clientVersion: opts.clientVersion,
    transport,
    baseUrl,
    proxyOrigin,
    caseStorage: opts.caseStorage,
  };
}

function engineBaseUrl(engine: IsaEngine): string {
  switch (engine.kind) {
    case 'remote':
    case 'local':
      return engine.baseUrl;
    case 'proxy':
      return PRODUCTION_REMOTE_ORIGIN;
    case 'in_memory':
      return PRODUCTION_REMOTE_ORIGIN;
  }
}
